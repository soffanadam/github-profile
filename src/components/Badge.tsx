import React from 'react'

export type BadgeProps = {
  dataTestID?: string
}

export const Badge: React.FC<BadgeProps> = ({ dataTestID, children }) => {
  return (
    <span
      data-testid={dataTestID}
      className="inline-flex items-center py-0.5 px-2.5 text-gray-500 bg-white rounded border"
    >
      {children}
    </span>
  )
}
