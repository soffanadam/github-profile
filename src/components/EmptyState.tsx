import { classNames } from '@/utils'
import React from 'react'

export type EmptyStateProps = {
  dataTestID?: string
  className?: string
}

export const EmptyState: React.FC<EmptyStateProps> = ({
  children,
  dataTestID,
  className
}) => {
  return (
    <div
      data-testid={dataTestID}
      className={classNames('py-3 px-5 text-xl', className)}
    >
      {children}
    </div>
  )
}
