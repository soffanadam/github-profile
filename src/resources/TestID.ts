export enum TestID {
  SEARCH_FORM = 'search-form',
  SEARCH_FORM_INPUT = 'search-form-input'
}
