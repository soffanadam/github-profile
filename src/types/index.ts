// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface PlainObject<T = any> {
  [key: string]: T
}

//==============================================================================
// Model
//==============================================================================

export interface User {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [anyProp: string]: any
}

export interface Repo {
  id: number
  name: string
  description?: string
  stargazers_count: number
  forks_count: number
  owner: User
}

export interface Content {
  name: string
  path: string
  type: string
  html_url?: string
}

export interface ReadMe {
  name: string
  content: string
  encoding: string
}

//==============================================================================
// State
//==============================================================================

export interface UserState {
  user: User | null
  error: string
  loading: boolean
  searching: boolean
}

export interface ReposState {
  repos: Repo[]
  page: number
  hasMore: boolean
  error: string
  loading: boolean
}

export interface RepoState {
  repo: Repo | null
  error: string
  loading: boolean
}

export interface ContentsState {
  path: string
  contents: Content[]
  readMe: ReadMe | null
  error: string
  loading: boolean
}

export interface RootState {
  userState: UserState
  reposState: ReposState
  repoState: RepoState
  contentsState: ContentsState
}

//==============================================================================
// Actions Payload
//==============================================================================

export interface GetRepoPayload {
  userName: string
  repoName: string
}

//==============================================================================
// API
//==============================================================================

export class ApiError extends Error {
  response: ApiErrorResponse | null

  constructor(message: string, response: ApiErrorResponse | null = null) {
    super(message)
    this.response = response
  }
}

export interface ApiErrorResponse {
  status?: number
  body?: unknown
}
