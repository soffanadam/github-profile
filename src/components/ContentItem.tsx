import { Content, ContentType } from '@/types'
import React from 'react'
import { FcFile, FcFolder, FcLink, FcPackage } from 'react-icons/fc'

export const ICON_TEST_ID = 'icon'

export type ContentItemProps = {
  dataTestID?: string
  content: Content
}

export const ContentItem: React.FC<ContentItemProps> = ({
  dataTestID,
  content
}) => {
  return (
    <span
      data-testid={dataTestID}
      className="flex items-center py-2 px-5 hover:bg-gray-50 border-b"
    >
      <span className="w-8">
        {content.type == ContentType.DIR ? (
          <FcFolder data-icon={ContentType.DIR} />
        ) : content.type == ContentType.SYMLINK ? (
          <FcLink data-icon={ContentType.SYMLINK} />
        ) : content.type == ContentType.SUBMODULE ? (
          <FcPackage data-icon={ContentType.SUBMODULE} />
        ) : (
          <FcFile data-icon={ContentType.FILE} />
        )}
      </span>
      {content.name}
    </span>
  )
}
