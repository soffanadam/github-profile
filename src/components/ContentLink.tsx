import { Content, ContentType, PlainObject } from '@/types'
import React from 'react'
import { Link } from 'react-router-dom'

export type ContentLinkProps = {
  dataTestID?: string
  repoUrl: string
  content: Content
}

export const ContentLink: React.FC<ContentLinkProps & PlainObject> = ({
  dataTestID,
  repoUrl,
  content,
  children,
  ...rest
}) => {
  if (content.type === ContentType.DIR) {
    return (
      <Link
        data-testid={dataTestID}
        to={`${repoUrl}/${content.path}`}
        {...rest}
      >
        {children}
      </Link>
    )
  }

  if (content.html_url) {
    return (
      <a
        data-testid={dataTestID}
        href={content.html_url}
        target="_blank"
        rel="noreferrer"
        {...rest}
      >
        {children}
      </a>
    )
  }

  return (
    <div data-testid={dataTestID} {...rest}>
      {children}
    </div>
  )
}
