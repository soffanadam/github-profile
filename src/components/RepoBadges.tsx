import { Repo } from '@/types'
import React from 'react'
import { FiStar, FiGitBranch } from 'react-icons/fi'
import { Badge } from './Badge'

export type RepoBadgesProps = {
  repo: Repo
}

export const RepoBadges: React.FC<RepoBadgesProps> = ({ repo }) => {
  return (
    <div className="flex mt-2 space-x-3">
      <Badge>
        <FiStar className="mr-2" />
        <span data-stargazers-count>{repo.stargazers_count}</span>
      </Badge>
      <Badge>
        <FiGitBranch className="mr-2" />
        <span data-forks-count>{repo.forks_count}</span>
      </Badge>
    </div>
  )
}
