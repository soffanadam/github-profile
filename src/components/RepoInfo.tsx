import { Repo } from '@/types'
import React from 'react'
import { Link } from 'react-router-dom'

export type RepoInfoProps = {
  dataTestID?: string
  repo: Repo
}

export const RepoInfo: React.FC<RepoInfoProps> = ({ dataTestID, repo }) => {
  return (
    <Link
      data-testid={dataTestID}
      to={`/${repo.owner.login}/${repo.name}`}
      className="group block"
    >
      <span
        data-testid="repo-info-name"
        className="inline-flex pb-1 mb-2 text-2xl text-indigo-500 border-b border-transparent group-hover:border-indigo-300"
      >
        {repo.name}
      </span>
      {repo.description && (
        <p data-testid="repo-info-description" className="mb-5">
          {repo.description}
        </p>
      )}
    </Link>
  )
}
