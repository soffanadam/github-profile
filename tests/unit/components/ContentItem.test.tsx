import React from 'react'
import { render } from '@testing-library/react'
import { ContentItem, ContentItemProps } from '@/components/ContentItem'
import { TestID } from '@/resources/TestID'
import { ContentType } from '@/types'
import { withContentItemProps } from './utils'

const renderComponent = (props: ContentItemProps) => {
  return render(<ContentItem {...props} />)
}

describe('<ContentItem />', () => {
  it('renders correctly', () => {
    const props = withContentItemProps(ContentType.DIR)
    const component = renderComponent(withContentItemProps(ContentType.DIR))
    expect(component).toBeTruthy()

    const testElement = component.getByTestId(TestID.CONTENT_ITEM)
    expect(testElement).toHaveTextContent(props.content.name)
    expect(testElement).toMatchSnapshot()
  })

  it('is a directory', () => {
    const { container } = renderComponent(withContentItemProps(ContentType.DIR))
    expect(
      container.querySelector(`[data-icon=${ContentType.DIR}]`)
    ).toBeVisible()
  })

  it('is a file', () => {
    const { container } = renderComponent(
      withContentItemProps(ContentType.FILE)
    )
    expect(
      container.querySelector(`[data-icon=${ContentType.FILE}]`)
    ).toBeVisible()
  })

  it('is a symlink', () => {
    const { container } = renderComponent(
      withContentItemProps(ContentType.SYMLINK)
    )
    expect(
      container.querySelector(`[data-icon=${ContentType.SYMLINK}]`)
    ).toBeVisible()
  })

  it('is a submodule', () => {
    const { container } = renderComponent(
      withContentItemProps(ContentType.SUBMODULE)
    )
    expect(
      container.querySelector(`[data-icon=${ContentType.SUBMODULE}]`)
    ).toBeVisible()
  })
})
