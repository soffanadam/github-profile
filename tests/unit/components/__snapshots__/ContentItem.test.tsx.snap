// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentItem /> renders correctly 1`] = `
<span
  class="flex items-center py-2 px-5 hover:bg-gray-50 border-b"
  data-testid="content-item"
>
  <span
    class="w-8"
  >
    <svg
      data-icon="dir"
      enable-background="new 0 0 48 48"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      version="1"
      viewBox="0 0 48 48"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M40,12H22l-4-4H8c-2.2,0-4,1.8-4,4v8h40v-4C44,13.8,42.2,12,40,12z"
        fill="#FFA000"
      />
      <path
        d="M40,12H8c-2.2,0-4,1.8-4,4v20c0,2.2,1.8,4,4,4h32c2.2,0,4-1.8,4-4V16C44,13.8,42.2,12,40,12z"
        fill="#FFCA28"
      />
    </svg>
  </span>
  contentname
</span>
`;
